code/modules/power/prototype_shield_emitter.dm

```diff
@@ -0,0 +1,203 @@
+		return
+	if(a.power_equip == FALSE)
+		to_chat(user, "<span class='warning'>There is no power in this area!!</span>")
+		return
+	if(locked)
+		to_chat(user, "<span class='warning'>The controls are locked!</span>")
+		return
+	if(is_on == TRUE)
+		remove_barrier(user)
+		return
+	check_map_borders(1,4,3,4)
+	if(!borders)
+		to_chat(user, "<span class='warning'>The motors whir and fail!</span>")
+		return
+	build_barrier(0,2,2,3,1,1,3,4,user)
+
+///Build the barriers
+/obj/machinery/power/proto_sh_emitter/proc/build_barrier(A,B,C,D,E,F,G,H,user)
+	var/turf/EmitterTurf = get_turf(src)
+	to_chat(user, "<span class='warning'>You start to turn on the [src] and the generated shields!</span>")
+	if(do_after(user, 1.5 SECONDS, target = src))
+		to_chat(user, "<span class='warning'>You turn on the [src] and the generated shields!</span>")
+		message_admins("[src] turned on at [ADMIN_VERBOSEJMP(EmitterTurf)] by [ADMIN_LOOKUPFLW(user)]")
+		log_game("[src] turned on at [AREACOORD(EmitterTurf)] by [key_name(user)]")
+		is_on = TRUE
+		update_icon_state()
+		switch(dir) //this part check the direction of the machine and create the block in front of it
+			if(NORTH)
+				LAZYADD(internal, block(locate(x - A, y + B, z), locate(x + C, y + D, z)))
+				LAZYADD(outline, block(locate(x - E, y + F, z), locate(x + G, y + H, z)) - internal)
+			if(SOUTH)
+				LAZYADD(internal, block(locate(x - C, y - B, z), locate(x + A, y - D, z)))
+				LAZYADD(outline, block(locate(x - G, y - F, z), locate(x + E, y - H, z)) - internal)
+			if(EAST)
+				LAZYADD(internal, block(locate(x + B, y - C, z), locate(x + D, y + A, z)))
+				LAZYADD(outline, block(locate(x + F, y - G, z), locate(x + H, y + E, z)) - internal)
+			if(WEST)
+				LAZYADD(internal, block(locate(x - B, y - A, z), locate(x - D, y + C, z)))
+				LAZYADD(outline, block(locate(x - F, y - E, z), locate(x - H, y + G, z)) - internal)
+		for(var/turf in outline)
+			new /obj/machinery/holosign/barrier/power_shield/wall(turf, src)
+			LAZYREMOVE(outline, turf)
+		for(var/turf in internal)
+			new /obj/machinery/holosign/barrier/power_shield/floor(turf, src)
+			LAZYREMOVE(internal, turf)
+
+///Remove the barriers
+/obj/machinery/power/proto_sh_emitter/proc/remove_barrier(user)
+	var/turf/EmitterTurf = get_turf(src)
+	to_chat(user, "<span class='warning'>You start to turn off the [src] and the generated shields!</span>")
+	if(do_after(user, 3.5 SECONDS, target = src))
+		to_chat(user, "<span class='warning'>You turn off the [src] and the generated shields!</span>")
+		message_admins("[src] turned off at [ADMIN_VERBOSEJMP(EmitterTurf)] by [ADMIN_LOOKUPFLW(user)]")
+		log_game("[src] turned off at [AREACOORD(EmitterTurf)] by [key_name(user)]")
+		is_on = FALSE
+		for(var/h in signs)
+			qdel(h)
+		update_icon_state()
+
+///Check if the machine is generating the barriers inside the map borders
+/obj/machinery/power/proto_sh_emitter/proc/check_map_borders(A,B,C,D)
```

Comment:
Can you set these vars to be more descriptive? What is `A`, what is `B`?
How about `distance_to_east_corner` and the likes?

Reply:
i'll set them as someone is looking north and say NEx NEy for north east x and y and so on, i'll do the same for the other proc, it would be better done like that because otherwise the naming would be so long. i'll put a proper comment on the proc too to explain how it works